packages/front-end/components/OpenVisualEditorLink.tsx

```diff
@@ -45,49 +61,81 @@ const OpenVisualEditorLink: FC<{
     });
 
     return url;
-  }, [visualEditorUrl, id, changeIndex, apiHost]);
+  }, [apiHost, changeIndex, id, visualEditorUrl]);
 
-  const navigate = useCallback(() => {
-    track("Open visual editor", {
-      source: "visual-editor-ui",
-      status: "success",
+  const getVisualEditorKey = useCallback(async () => {
+    const res = await apiCall<{ key: string }>("/visual-editor/key", {
+      method: "GET",
     });
-    window.location.href = url;
+    return res.key;
+  }, [apiCall]);
+
+  const navigate = useCallback(async () => {
+    setShowExtensionDialog(false);
+
+    setIsNavigating(true);
+
+    const key = await getVisualEditorKey();
+
+    window.postMessage(
+      {
+        type: "GB_REQUEST_OPEN_VISUAL_EDITOR",
+        data: key,
+      },
+      window.location.origin
+    );
+
+    // in the case a user has clicked 'proceed anyway' when we do not detect the
+    // chrome extension installation, we ignore waiting for the responsem msg
+    // and navigate right away.
+    if (isBypassing) {
+      setIsNavigating(false);
+
+      track("Open visual editor", {
+        source: "visual-editor-ui",
+        status: "success",
+      });
+
+      window.location.href = url;
+    }
+  }, [url, getVisualEditorKey, isBypassing]);
+
+  // we wait until the visual editor gives us a response message to open a new
+  // window. this ensures that the api key is set upon loading it.
+  useEffect(() => {
+    if (!url) return;
+
+    const onMessage = (
+      event: MessageEvent<{ type?: "GB_RESPONSE_OPEN_VISUAL_EDITOR" }>
+    ) => {
+      if (event.data.type === "GB_RESPONSE_OPEN_VISUAL_EDITOR") {
+        track("Open visual editor", {
+          source: "visual-editor-ui",
+          status: "success",
+        });
+
+        setIsNavigating(false);
+
+        window.location.href = url;
+      }
+    };
+
+    window.addEventListener("message", onMessage);
+
+    return () => window.removeEventListener("message", onMessage);
   }, [url]);
 
   return (
     <>
       <span
         className="btn btn-sm btn-primary"
+        style={{ width: "144px" }}
         onClick={async (e) => {
-          if (!visualEditorUrl) {
-            e.preventDefault();
-            setShowEditorUrlDialog(true);
-            track("Open visual editor", {
-              source: "visual-editor-ui",
-              status: "missing visualEditorUrl",
-            });
-            return false;
-          }
+          e.preventDefault();
```

Comment:
If `isNavigating` is true, do we want to make `onClick` do nothing?  Right now, it looks like it would fire duplicate `navigate` calls.